### PR TITLE
Tutorial post on system configuration

### PR DESCRIPTION
**Pull request by gracecodes144**
A bash script to get CentOS 6 machines up and running with system
packages needed for Harvest and code contribution, including python 2.7

I walked two people through various stages of system configuration
BEFORE installing harvest and starting a project, so I thought I'd combine
that experience with puppet scripts by @tjrivera and @wraziens into a
simple bash script users who aren't into puppet, vagrant, etc can run just
to get started. This is what I came up with.

### CODE REVIEW

File: _posts/2014-02-09-system-config.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+```
+
+### Compile and install python 2.7
+
+```bash
+cd /home/devuser/downloads
+wget http://www.python.org/ftp/python/2.7.3/Python-2.7.3.tgz
+tar -xzf Python-2.7.3.tgz
+cd Python-2.7.3
+/bin/bash configure --prefix=/home/devuser/local
```
Comment (bruth) [inline _posts/2014-02-09-system-config.md]: Add the `--enable-shared` flag as well. Allows for shared dynamic linking of libraries.

File: _posts/2014-02-09-system-config.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+cd /home/devuser/downloads 
+wget https://raw.github.com/pypa/pip/master/contrib/get-pip.py
+python get-pip.py
+pip install distribute virtualenv uwsgi supervisor
+```
+
+### Install node.js, npm, and important node packages
+
+```bash
+wget http://nodejs.org/dist/node-latest.tar.gz && tar -xzf node-latest.tar.gz
```
Comment (bruth) [inline _posts/2014-02-09-system-config.md]: If you want a one-liner, `curl "http://nodejs.org/dist/node-latest.tar.gz" | tar xz`

File: _posts/2014-02-09-system-config.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+```
+
+At this point, executing `which python` should return `/home/devuser/local/bin/python`. If it doesn't, try reloading your shell session.
+
+### Install pip and important python packages
+
+```bash
+cd /home/devuser/downloads 
+wget https://raw.github.com/pypa/pip/master/contrib/get-pip.py
+python get-pip.py
```
Comment (bruth) [inline _posts/2014-02-09-system-config.md:80]: `curl "https://raw.github.com/pypa/pip/master/contrib/get-pip.py" | python`

File: _posts/2014-02-09-system-config.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+### Create `/home/devuser/local` and configure PATH
+
+I'll break caveat 1 here to say that I chose `/home/devuser/local` as the install prefix so that if a sysadmin was able to run the above portion, the rest of the script can be run without root privileges of any kind (unlike installing to `/opt` or `/usr/local`). It is also extremely unlikely to conflict with other users or applications on your system (as installing to `/usr/local` might).
+
+```bash
+mkdir /home/devuser/local
+cd
+echo 'export PATH=/home/devuser/local/bin:${PATH}' >> .bashrc
+source .bashrc
```
Comment (bruth) [inline _posts/2014-02-09-system-config.md]: No need for the `cd`

```
mkdir /home/devuser/local
echo 'export PATH=/home/devuser/local/bin:${PATH}' >> ~/.bashrc
source ~/.bashrc
```

File: _posts/2014-02-09-system-config.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+mkdir downloads
+cd downloads
+wget http://dl.fedoraproject.org/pub/epel/6/x86_64/epel-release-6-8.noarch.rpm
+sudo rpm -Uvh ./epel-release-6-8.noarch.rpm
+```
+
+### Install system packages and then upgrade everything
+
+```bash
+sudo yum install httpd-devel openssl-devel ncurses-devel rpm-devel apr-util-devel apr-devel glibc-devel memcached readline-devel bzip2 bzip2-devel bzip2-libs libpng-devel openldap-devel freetype fontconfig freetype-devel lapack-devel blas-devel libgfortran gcc-gfortran gcc 'gcc-c++' postgresql-devel zlib-devel libcurl-devel expat-devel gettext-devel nginx sqlite-devel vim python-devel gmp-devel man
```
Comment (bruth) [inline _posts/2014-02-09-system-config.md:44]: Not sure if it's necessary, but some of these libraries are optional depending on the web server and database choice. Maybe all you need to state is that these include the libraries for a production deployment as well.

File: _posts/2014-02-09-system-config.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+Two caveats:
+
+1. These steps represent one way of doing things, not the way. This guide intentionally eschews justification in deference to brevity.
+2. These steps are intended for CentOS 6 (64 bit arch) because that's the OS in which I do development. With minor modification, they would probably work on Redhat. With major modification, but similar intent, they would probably work on most Linux distros. Again in deference to brevity, I didn't try for universality.
+
+### Make a user for shared app development
+
+```bash
+sudo useradd -m devuser
+sudo chmod 770 /home/devuser
```
Comment (bruth) [inline _posts/2014-02-09-system-config.md]: Add this line as well `sudo chmod g+s /home/devuser` which ensures the group (`devuser`) is maintained for all newly created subdirectories. This prevents the deploying user's group from polluting all the subdirectories.

Comment (murphyke) [inline _posts/2014-02-09-system-config.md]: Or `sudo chmod 2770 /home/devuser`, which includes the setgid bit in the original chmod.

**Comment by gracecodes144** [comment]
@bruth Incorporated all your suggestions except for the one-liners as I feel for less experience users that can be confusing, and if a more experienced user does come into the system afterward, at least there will be evidence of the install methods.

**Comment by gracecodes144** [comment]
re: devuser home, would `sudo chmod g+srw /home/devuser` work?

**Comment by gracecodes144** [comment]
except with the execute bit as well, obviously...

**Comment by gracecodes144** [comment]
tested in vagrant, `sudo chmod g+rwxs /home/devuser` wins, for compactness combined with clarity

**Comment by gracecodes144** [comment]
@bruth @murphyke ready for dev so others can see it in their browsers?

**Comment by murphyke** [comment]
I trust you if you have run through the script.  One thing; Python is now at 2.7.6.  You could add a note about checking for the latest 2.7 release.

**Comment by gracecodes144** [comment]
@murphyke good point, note added. I ran the script through the whole way in a vagrant box, so I know it works. And what I kept hearing from people last week was that they understand every system is different, but they need somewhere to start.... and honestly it will be much easier to assist people if they come and say, I ran your script and it failed here, than if they say, I don't know what to do?

**Pull request merged**